File: common/src/main/java/wile/redstonepen/blocks/controlbox/ControlBoxUiContainer.java
package wile.redstonepen.blocks.controlbox;

import net.minecraft.nbt.CompoundTag;
import net.minecraft.network.chat.Component;
import net.minecraft.world.Container;
import net.minecraft.world.SimpleContainer;
import net.minecraft.world.entity.player.Inventory;
import net.minecraft.world.entity.player.Player;
import net.minecraft.world.inventory.*;
import net.minecraft.world.item.ItemStack;
import net.minecraft.world.level.Level;
import wile.redstonepen.libmc.Networking;
import wile.redstonepen.libmc.Registries;

import org.jetbrains.annotations.Nullable;


@SuppressWarnings("deprecation")
public class ControlBoxUiContainer extends AbstractContainerMenu implements Networking.INetworkSynchronisableContainer
{
  protected static final int NUM_OF_SLOTS = 1;
  protected final Player player_;
  protected final Container inventory_;
  protected final ContainerLevelAccess wpc_;
  private final ContainerData fields_;
  private volatile CompoundTag received_server_data_ = new CompoundTag();
  //------------------------------------------------------------------------------------------------------------------
  public int field(int index) { return fields_.get(index); }
  public Player player() { return player_ ; }
  public Container inventory() { return inventory_ ; }
  public Level world() { return player_.level(); }
  public @Nullable ControlBoxBlockEntity te() { return wpc_.evaluate((w,p)->{net.minecraft.world.level.block.entity.BlockEntity te=w.getBlockEntity(p); return (te instanceof ControlBoxBlockEntity cbte) ? (cbte): (null); }).orElse(null); }
  //------------------------------------------------------------------------------------------------------------------

  public ControlBoxUiContainer(int cid, Inventory player_inventory)
  { this(cid, player_inventory, new SimpleContainer(ControlBoxUiContainer.NUM_OF_SLOTS), ContainerLevelAccess.NULL, new SimpleContainerData(1)); }

  public ControlBoxUiContainer(int cid, Inventory player_inventory, Container block_inventory, ContainerLevelAccess wpc, ContainerData fields)
  {
    super(Registries.getMenuTypeOfBlock("control_box"), cid);
    player_ = player_inventory.player;
    inventory_ = block_inventory;
    wpc_ = wpc;
    wpc_.execute((w,p)->inventory_.startOpen(player_));
    fields_ = fields;
    addDataSlots(fields_);
    for(int x=0; x<9; ++x) addSlot(new Slot(player_inventory, x, 28+x*18, 183)); // player hotbar slots: 0..8
  }

  @Override
  public boolean stillValid(Player player)
  { return inventory_.stillValid(player); }

  @Override
  public void removed(Player player)
  { super.removed(player); inventory_.stopOpen(player); }

  @Override
  public void sendAllDataToRemote()
  {
    super.sendAllDataToRemote();
    if((world().isClientSide) || (te()==null)) return;
    Networking.PacketContainerSyncServerToClient.sendToListeners(world(), this, composeServerData(te(), true));
  }

  @Override
  public ItemStack quickMoveStack(Player player, int slot)
  { return ItemStack.EMPTY; }

  // Container client/server synchronization --------------------------------------------------

  public CompoundTag composeServerData(ControlBoxBlockEntity te, boolean full)
  { return te.collectSyncData(full); }

  public CompoundTag fetchReceivedServerData()
  {
    final CompoundTag received = received_server_data_;
    received_server_data_ = new CompoundTag();
    return received;
  }

  @Override
  public void onServerPacketReceived(int windowId, CompoundTag nbt)
  {
    switch(nbt.getString("action")) {
      case "serverdata" -> { received_server_data_ = nbt; }
      default -> {}
    }
  }

  @Override
  public void onClientPacketReceived(int windowId, Player player, CompoundTag nbt)
  {
    final ControlBoxBlockEntity te = te();
    if(te==null) return;
    int sync = 0;
    switch(nbt.getString("action")) {
      case "codeupdate" -> { te.setCode(nbt.getString("code")); }
      case "serverdata" -> { sync = 2; }
      case "servervalues" -> { sync = 1; }
      case "enabled" -> {
        te.setEnabled(!te.getEnabled());
        te.setRcaPlayerUUID((te.getEnabled() && nbt.getBoolean("withrca")) ? player.getUUID() : null);
        sync = 2;
      }
      default -> {
      }
    }
    if(sync > 0) Networking.PacketContainerSyncServerToClient.sendToListeners(world(), this, composeServerData(te, sync>1));
  }
}

File: common/src/main/kotlin/wile/redstonepen/blocks/controlbox/ControlBoxUiContainer.kt
package wile.redstonepen.blocks.controlbox

import net.minecraft.nbt.CompoundTag
import net.minecraft.world.Container
import net.minecraft.world.SimpleContainer
import net.minecraft.world.entity.player.Inventory
import net.minecraft.world.entity.player.Player
import net.minecraft.world.inventory.AbstractContainerMenu
import net.minecraft.world.inventory.ContainerData
import net.minecraft.world.inventory.ContainerLevelAccess
import net.minecraft.world.inventory.SimpleContainerData
import net.minecraft.world.inventory.Slot
import net.minecraft.world.item.ItemStack
import net.minecraft.world.level.Level
import wile.redstonepen.net.Networking
import wile.redstonepen.registry.Registries

@Suppress("DEPRECATION")
class ControlBoxUiContainer(
    cid: Int,
    playerInventory: Inventory,
    blockInventory: Container,
    wpc: ContainerLevelAccess,
    fields: ContainerData,
) :
    AbstractContainerMenu(Registries.getMenuTypeOfBlock("control_box"), cid),
    Networking.INetworkSynchronisableContainer {

    companion object {
        private const val NUM_OF_SLOTS = 1
    }

    constructor(
        cid: Int,
        playerInventory: Inventory,
    ) : this(
        cid,
        playerInventory,
        SimpleContainer(NUM_OF_SLOTS),
        ContainerLevelAccess.NULL,
        SimpleContainerData(1),
    )

    private val player: Player = playerInventory.player
    private val inventory: Container = blockInventory
    private val wpc: ContainerLevelAccess = wpc
    private val fields: ContainerData = fields

    @Volatile private var receivedServerData: CompoundTag = CompoundTag()

    init {
        this.wpc.execute { _, _ -> this.inventory.startOpen(this.player) }
        addDataSlots(this.fields)
        for (x in 0 until 9) addSlot(Slot(playerInventory, x, 28 + x * 18, 183))
    }

    fun field(index: Int): Int = fields.get(index)

    fun player(): Player = player

    fun inventory(): Container = inventory

    fun world(): Level = player.level()

    fun te(): ControlBoxBlockEntity? =
        wpc.evaluate { w, p -> w.getBlockEntity(p) as? ControlBoxBlockEntity }.orElse(null)

    override fun stillValid(player: Player): Boolean = inventory.stillValid(player)

    override fun removed(player: Player) {
        super.removed(player)
        inventory.stopOpen(player)
    }

    override fun sendAllDataToRemote() {
        super.sendAllDataToRemote()
        if (world().isClientSide) return
        val te = te() ?: return
        Networking.PacketContainerSyncServerToClient.sendToListeners(
            world(),
            this,
            composeServerData(te, true),
        )
    }

    override fun quickMoveStack(player: Player, slot: Int): ItemStack = ItemStack.EMPTY

    fun composeServerData(te: ControlBoxBlockEntity, full: Boolean): CompoundTag =
        te.collectSyncData(full)

    fun fetchReceivedServerData(): CompoundTag {
        val received = receivedServerData
        receivedServerData = CompoundTag()
        return received
    }

    override fun onServerPacketReceived(windowId: Int, nbt: CompoundTag) {
        when (nbt.getString("action")) {
            "serverdata" -> {
                receivedServerData = nbt
            }
        }
    }

    override fun onClientPacketReceived(windowId: Int, player: Player, nbt: CompoundTag) {
        val te = te() ?: return
        var sync = 0
        when (nbt.getString("action")) {
            "codeupdate" -> {
                te.setCode(nbt.getString("code"))
            }
            "serverdata" -> {
                sync = 2
            }
            "servervalues" -> {
                sync = 1
            }
            "enabled" -> {
                te.setEnabled(!te.getEnabled())
                te.setRcaPlayerUUID(
                    if (te.getEnabled() && nbt.getBoolean("withrca")) player.uuid else null
                )
                sync = 2
            }
        }
        if (sync > 0) {
            Networking.PacketContainerSyncServerToClient.sendToListeners(
                world(),
                this,
                composeServerData(te, sync > 1),
            )
        }
    }
}
